library/file.lisp

```diff
@@ -0,0 +1,663 @@
+                                                  (IfExists/Supersede ':supersede))
+                                     :if-does-not-exist (cl:case if-does-not-exist
+                                                          (IfDoesNotExist/DNExistError ':error)
+                                                          (IfDoesNotExist/Create ':create))))))
+
+  (declare %open-twoway (Pathname -> IFExists -> IfDoesNotExist -> types:lisptype -> (Result FileError (FileStream :a))))
+  (define (%open-twoway path if-exists if-does-not-exist etype)
+    "Opens a two way stream for the given filepath and for a given type."
+    (lisp (Result FileError (FileStream :a)) (path if-exists if-does-not-exist etype)
+      (handle-file-function (cl:open path
+                                     :direction :io
+                                     :element-type etype
+                                     :if-exists (cl:case if-exists
+                                                  (IfExists/ExistsError ':error)
+                                                  (IfExists/NewVersion ':new-version)
+                                                  (IfExists/Rename ':rename)
+                                                  (IfExists/RenameAndDelete ':rename-and-delete)
+                                                  (IfExists/Overwrite ':overwrite)
+                                                  (IfExists/Append ':append)
+                                                  (IfExists/Supersede ':supersede))
+                                     :if-does-not-exist (cl:case if-does-not-exist
+                                                          (IfDoesNotExist/DNExistError ':error)
+                                                          (IfDoesNotExist/Create ':create))))))
+
+  (declare %open (StreamOptions -> types:lisptype -> (Result FileError (FileStream :a))))
+  (define (%open stream-options etype)
+    "Opens a FileStream for a given type and StreamOptions."
+    (match stream-options
+      ((Input path)
+       (%open-input path etype))
+      ((Output path exists does-not-exist)
+       (%open-output path exists does-not-exist etype))
+      ((TwoWay path exists does-not-exist)
+       (%open-twoway path exists does-not-exist etype))))
+
+  ;;
+  ;; Other basic stream operations
+  ;;
+  (declare %close ((FileStream :a) -> (Result FileError :b)))
+  (define (%close stream)
+    "Closes a FileStream."
+    (lisp (Result FileError :a) (stream)
+      (handle-file-function (cl:close stream))))
+
+  (declare read-char ((FileStream Char) -> (Result FileError Char)))
+  (define (read-char stream)
+    "Reads a character from an FileStream."
+    (lisp (Result FileError Char) (stream)
+      (cl:handler-case (Ok (cl:read-char stream))
+        (cl:end-of-file () (Err (EOF)))
+        (cl:error (c) (Err (LispError c))))))
+
+  (declare read-byte ((FileStream :a) -> (Result FileError :a)))
```

Comment:
`Result FileError Byte`? also should this be allowed on char streams?

Reply:
Does Byte include all integer subtypes? This works for All types `U8`, `I8`...`UFix` `IFix`.

Reply:
Oh I can use the constraint `Bounded :a`!